### ISSUE TITLE
Why to you detach fake_H?

### ISSUE THREAD

**Issue by andreas128** [xinntao/BasicSR#17]
Why do you detach the fake_H in `SRRaGAN_model.py`? The update would not be applied to G, as you did not add the G parameter to the optimizer. Is that correct?

`pred_d_real = self.netD(self.var_ref)`
`pred_d_fake = self.netD(self.fake_H.detach())  # detach to avoid BP to G`

optimizer:
`self.optimizer_D = torch.optim.Adam(self.netD.parameters(), lr=train_opt['lr_D'], \
                weight_decay=wd_D, betas=(train_opt['beta1_D'], 0.999))`

**Comment by xinntao** [comment]
Hi, @andreas128 
I think the code is correct. 

The procedure is that we train netG and netD iteratively. 
The code you paste is for training the netD. When we train the netD, we only update the parameters of the netD. We detach the `self.fake_H` so that the gradients will not back-propagate the the netG. (If we omit detach operation, then the gradients will also back-propagate to the netG. But it is useless, since in this step we only need to update the netD)

The update for the netG is in  [other lines](https://github.com/xinntao/BasicSR/blob/71635757505ef113bfba029441a4e35e59994a86/codes/models/SRRaGAN_model.py#L133):
`pred_g_fake = self.netD(self.fake_H)`, in this line, the gradients will back-propagate to the self.fake_H and then the netG. The optimizer `self.optimizer_G` will update the G parameters.
